lib/modules/manager/dockerfile/extract.ts

```diff
@@ -11,6 +11,70 @@ const variableOpen = '${';
 const variableClose = '}';
 const variableDefaultValueSplit = ':-';
 
+export function extractVariables(image: string): Record<string, string> {
+  const variables: Record<string, string> = {};
+  const variableRegex = regEx(
+    /(?<fullvariable>\\?\$(?<simplearg>\w+)|\\?\${(?<complexarg>\w+)(?::.+?)?}+)/gi
+  );
+
+  let match: RegExpExecArray | null;
+  do {
+    match = variableRegex.exec(image);
+    if (match?.groups?.fullvariable) {
+      variables[match.groups.fullvariable] =
+        match.groups?.simplearg || match.groups?.complexarg;
+    }
+  } while (match);
+
+  return variables;
+}
+
+function processDepForAutoReplace(
+  dep: PackageDependency,
+  lineNumberRanges: number[][],
+  lines: string[],
+  linefeed: string
+): string | undefined {
+  const lineNumberRangesToReplace: number[][] = [];
+  for (const lineNumberRange of lineNumberRanges) {
+    for (const lineNumber of lineNumberRange) {
+      if (
+        (dep.currentValue && lines[lineNumber].includes(dep.currentValue)) ||
```

Comment:
is there a chance that when you try to find `dep.currentValue` it could cause a mistake if the keyword exists in multiple lines? isn't it better to use regex to find the line number range like we do in other managers?

Reply:
`lineNumberRanges` is already confined to a subset of only those lines, which contribute to one specific FROM instruction. In theory, you're right that the keywords could occur in multiple lines but since an image definition specifies a distinctive version and digest only once, I don't think this is a real problem. However, if this ever happened, it could only be if FROM is used with `currentValue` and `currentDigest` spread across multiple ARG variables, because only then we would have multiple `lineNumberRange` entries. That again limits the likelihood of this event.

Later on in `extractPackageFile`, RegEx is used to identify the actual line ranges. The goal of the snippet you quoted does is to further narrow down what should be contained in `replaceString`. Imagine something like this:
```dockerfile
ARG ver=alpine:3.15.4
# now a bunch of other, completely unrelated instructions
# other ARG and FROM combinations for which there will be updates
#
# now later in the Dockerfile:
FROM $ver as stageX
```
Assume other dependency updates happen now in between. If we would take the block from `ARG ver...` to `FROM $ver...` as `replaceString`, there would be no `searchString` match in `auto-replace.ts` as `existingContent` is meanwhile different.

Reply:
That sounds like a really a long shot, guess we can keep it that way